akd/src/append_only_zks.rs

```diff
@@ -242,25 +230,28 @@ impl Azks {
                     .await?;
             load_count += nodes.len() as u64;
 
-            current_nodes = Vec::<NodeKey>::new();
-
-            // Now that states are loaded in the cache, you can read and access them.
-            // Note, the two for loops are needed because otherwise, you'd be accessing remote storage
+            // Now that states are loaded in the cache, we can read and access them.
+            // Note, we perform directional loads to avoid accessing remote storage
             // individually for each node's state.
-            for node in &nodes {
-                if !nodes_to_load.contains(&node.label) {
-                    // Only continue to traverse nodes which are relevant prefixes to insertion_set
-                    continue;
-                }
-
-                for dir in crate::DIRECTIONS {
-                    let child_label = node.get_child_label(dir)?;
-                    if let Some(child_label) = child_label {
-                        current_nodes.push(NodeKey(child_label));
-                    }
-                }
-            }
```

Comment:
Might be an unpopular opinion, but I actually find this imperative approach easier to understand than the functional one. I think the functional approach can be simplified further though (see clippy lints), maybe that would help readability :)

Reply:
Fair enough! I debated between the two, as I found the iteration over `DIRECTIONS` a bit "meh". I have incorporated the clippy lints as this point, and that does help a bit, so let me know what you think! I am happy to go with whatever the majority finds most readable on this one.

Reply:
Ah yes this is better! If we want to avoid writing the panic statement, we can also do:
```rust
current_nodes = nodes
    .iter()
    .filter(|node| filter_predicate(node))
    .flat_map(|node| {
        DIRECTIONS
            .iter()
            .filter_map(|dir| node.get_child_label(*dir).ok().flatten().map(NodeKey))
            .collect::<Vec<NodeKey>>()
    })
    .collect();
```
But if we want an explicit panic that's fine too 🙂

Reply:
I personally prefer the explicit panic, as I find it more explicit. That said, I'd like to get a confirmation from @kevinlewi around a panic at all, as this used to just result in an error. I questioned that, as `DIRECTIONS` is only left and right + the method we're calling is OK with left and right.

Reply:
Main thing is that a panic will result in a server crashing, right? The `get_child_label()` call could throw a StorageError due to storage being inaccessible for some reason, and that error would need to be handled

Reply:
Yes, that's right! I think the `get_child_label()` doesn't throw errors unless `Direction::None` is passed (see: https://github.com/facebook/akd/blob/main/akd/src/tree_node.rs#L503-L514). Since `DIRECTIONS` is composed of only `Left` and `Right`, it seemed like we could safely assume that this call should never result in an error at this callsite. Please do let me know if I am overlooking something!

Reply:
Oh sorry, I was looking at `get_child_node()` and not `get_child_label()`, my bad. I see what you mean.

Still, it feels like maybe the right thing to do would be to get rid of Direction::None, or create a DirectionWithoutNone kind of enum and have get_child_label() function take that as an argument, since get_child_label() should ideally not throw an error there either. But this is of course a much larger refactor...

Reply:
Perhaps we can put keep the panic for now, but add a TODO to remove it? I just am not a fan of having a possibility for the server to completely crash at run-time (as opposed to a catch-able error).

Reply:
@kevinlewi definitely down for that! I agree that this should be a compile-time error, if possible. I also agree that seems like a slightly larger refactor, so I am happy to tackle that in a separate change since it'll probably sprawl out a bit more.

If you're most comfortable with maintaining existing behavior as opposed to potentially introducing a server crash, I am fine with that as well. I honestly opted for the panic here to make the error more apparent in the event that it does happen considering that it really shouldn't and may be hard to find if it did.

Let me know what you think :D

Reply:
Sure, if you can file an issue to track this, and keep the expect there, that sounds good to me!

Reply:
@kevinlewi created https://github.com/facebook/akd/issues/314 to track the issue. I'll add a FIXME/TODO to this section of code to track the issue.